### DISCUSSION TITLE
Extracting test binaries

### DISCUSSION THREAD

**Discussion by TheNeikos** [ipetkov/crane#145] [Q&A] [answered]
I am currently writing some tests that require the docker daemon to work, hence I cannot run them inside the nix sandbox. My idea was to use `buildPackage` with `cargoExtraArgs = "--tests"` so that it would also build the tests and hopefully also extract them. Then I could run them manually from the outside.

Sadly, that did not work, the output derivation was simply empty. The next problem is also, that test binaries have a checksum in them, so I cannot hardcode extract them either.

Is there anything from the great toolset that crane provides that would get me close to that goal?

**Comment by ipetkov** [comment, marked as answer]
I would suggest adapting what `buildPackage` does for extracting the test binaries:

1. Run the `cargo test --no-run --message-format json-render-diagnostics` and capture the logs ([example](https://github.com/ipetkov/crane/blob/1bea2b52128832ab0e94370930e96ff0c99cf464/lib/buildPackage.nix#L42))
2. Use `jq` to filter the binaries you care about. `buildPackage` [ignores tests and only looks for libraries or other executables](https://github.com/ipetkov/crane/blob/1bea2b52128832ab0e94370930e96ff0c99cf464/lib/setupHooks/installFromCargoBuildLogHook.sh#L12-L14) but you might be able to achieve it with a filter like `'select(.reason == "compiler-artifact" and .profile.test == true) | .executable'` (you may need to tweak it further to suit your needs but from a quick test it looks like it will achieve what you are looking for)
3. The `jq` output will provide all the test binaries which would have been run for this test, and they can be copied/installed as appropriate

Hope this helps!

**Reply by TheNeikos** [reply]
That worked exactly as you wrote it down. Thank you very much!
